### ISSUE TITLE
split up into server/client

### ISSUE THREAD

**Issue by december-soul** [alf632/sandbox#9]
The client can be a RasPi which is connected to the kinect and projector.
The Server can be a powerful server (or even an AWS instance).

This will make it easy to build an low budget sandbox.

i have tried to user usbip to transfer the kinect to a server and use vnc to show the window on the client but the kinect use 3 usb devices bundeld into a virtual hub. usbip don't support it.

**Comment by ghost** [comment]
Streaming in a sufficiently high quality with a low delay is a very challenging job (see how poor Steam in-home streaming works with slower PCs or Wifi), and decoding video beyond very lowres compressed SD resolution without lag requires either a fast CPU or a dedicated hardware decoding unit with working drivers.

Then again our framerate is currently low anyway. However, VNC is most likely not up for the task if we manage to get it higher for the future, since there are too many changes to the image with people playing and VNC is generally performing pretty poorly with moving images with many changes.

The resulting latency since everything makes two trips over the connection (USB first, then the non-trivial encoding time, then the video returning) is also a potential problem.

Therefore, I think to get any decent performance it would require proper video streaming on the server of the output, and I'm not sure if it's ever going to work with usable latency unless the server is also in your local LAN (that is, not AWS).

I'm not super opposed to the idea, but it is certainly not going to be easy and whether it will ever work nicely with cloud servers far away is something I'm not even certain about.

Therefore, for now I would like to focus our time (since we're not working *that* much of our time on this project anyway) on more lowhanging fruit, like getting more stuff rendered on the GPU and more things in parallel threads, and adding some of the many features people want (kinect 2, lava, better possibly self-driving cars, digging water, nicer water with some rippling shader,...).

**Comment by december-soul** [comment]
you are right, full stream will have much latency. Maybe there is a point to send complex calculations to the server and just the result back to the client. Surface rendering can be done by the client. As i know rendering the surface is NOT the bottleneck. Waterflow is much more complex.

I have a server with an good GPU. Why not use this for water.

but anyways, i understand that there are more useful features.
can't wait to see driving cars :-)

**Comment by ghost** [comment]
Streaming the water has all the same challenges, since again it's a fullscreen image with lots of movement which requires low latency and higher framerate, and possibly a proper video stream (since VNC can't handle this nicely).
